[5.3] Fix schema.org with invalid breadcrumb json-ld

bembelimen opened the pull request:
Alternative for #45316

### Summary of Changes
This PR is a alternative fix for https://github.com/joomla/joomla-cms/issues/45277 where the connection is improved instead of deleted.


### Testing Instructions
1. Apply patch
2. Activate a breadcrumb module either on "none" or a non-existing position on your site


### Actual result BEFORE applying this Pull Request
- [Google](https://developers.google.com/search/docs/appearance/structured-data) reports invalid JSON-data


### Expected result AFTER applying this Pull Request
- JSON data is valid


### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [x] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [x] No documentation changes for manual.joomla.org needed

bembelimen committed:
Fix schema.org with missing breadcrumb display

bembelimen committed:
Merge branch '5.3-dev' into 5.3/schema-listitem

HLeithner commented:
I have tested this item :red_circle: unsuccessfully on 5715410d807979c2b7fff52fb18e119e4ed212e0<br /><br />Using this PR removes the link to the breadcrumbs json+ld entry if your template uses dynamic module positions which are not listed in the xml. Which removes the entry on my site because I use the breadcrumbs position from cassiopeia without defining it my self.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45720">issues.joomla.org/tracker/joomla-cms/45720</a>.</sub>

HLeithner committed:
Read breadcrumb ld+json id from the webassetmanager

Also rename the mod_bradcrumbs webasset name to a generic version which allow 3rd party extensions to specific their own implementation.

HLeithner committed:
cs

bembelimen commented on modules/mod_breadcrumbs/tmpl/default.php:
```diff
@@ -121,7 +121,7 @@
             'script',
             json_encode($data, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE | $prettyPrint),
-            ['name' => 'inline.mod_breadcrumbs-schemaorg'],
+            ['name' => 'inline.breadcrumb-schemaorg'],
```
```suggestion
            ['name' => 'inline.breadcrumbs-schemaorg'],
```

bembelimen commented on plugins/system/schemaorg/src/Extension/Schemaorg.php:
```diff
@@ -383,11 +387,30 @@ public function onBeforeCompileHead(): void
+        // Support Breadcrumb Schema linking
+        try {
+            try {
+                $breadcrumbAsset = $wa->getRegistry()->get('script', 'inline.breadcrumb-schemaorg');
```
```suggestion
                $breadcrumbsAsset = $wa->getRegistry()->get('script', 'inline.breadcrumbs-schemaorg');
```

bembelimen commented on plugins/system/schemaorg/src/Extension/Schemaorg.php:
```diff
@@ -383,11 +387,30 @@ public function onBeforeCompileHead(): void
+                $breadcrumbAsset = $wa->getRegistry()->get('script', 'inline.breadcrumb-schemaorg');
+            } catch (UnknownAssetException $e) {
+                // Fallback for older versions of the breadcrumbs module
+                $breadcrumbAsset = $wa->getRegistry()->get('script', 'inline.mod_breadcrumbs-schemaorg');
```
```suggestion
                $breadcrumbsAsset = $wa->getRegistry()->get('script', 'inline.mod_breadcrumbs-schemaorg');
```

bembelimen commented on plugins/system/schemaorg/src/Extension/Schemaorg.php:
```diff
@@ -383,11 +387,30 @@ public function onBeforeCompileHead(): void
+                trigger_deprecation(
+                    'joomla/schemaorg',
+                    '5.4',
+                    'The inline.mod_breadcrumbs-schemaorg asset name is deprecated. Please use the generic inline.breadcrumb-schemaorg asset name instead.'
```
```suggestion
                    'The inline.mod_breadcrumbs-schemaorg asset name is deprecated. Please use the generic inline.breadcrumbs-schemaorg asset name instead.'
```

bembelimen commented on plugins/system/schemaorg/src/Extension/Schemaorg.php:
```diff
@@ -383,11 +387,30 @@ public function onBeforeCompileHead(): void
+                );
+            }
+
+            $breadcrumb = json_decode($breadcrumbAsset->getOption('content'), true, 512, JSON_THROW_ON_ERROR);
```
```suggestion
            $breadcrumbs = json_decode($breadcrumbsAsset->getOption('content'), true, 512, JSON_THROW_ON_ERROR);
```

bembelimen commented on plugins/system/schemaorg/src/Extension/Schemaorg.php:
```diff
@@ -383,11 +387,30 @@ public function onBeforeCompileHead(): void
+            if ($breadcrumb['@type'] !== 'BreadcrumbList') {
+                trigger_error('The breadcrumb schema is not of type BreadcrumbList', E_USER_WARNING);
+                throw new UnknownAssetException();
+            }
```
```suggestion
            if ($breadcrumbs['@type'] !== 'BreadcrumbList') {
                trigger_error('The breadcrumbs schema is not of type BreadcrumbList', E_USER_WARNING);
                throw new UnknownAssetException();
            }
```

bembelimen commented on plugins/system/schemaorg/src/Extension/Schemaorg.php:
```diff
@@ -383,11 +387,30 @@ public function onBeforeCompileHead(): void
-            $webPageSchema['breadcrumb'] = ['@id' => $domain . '#/schema/BreadcrumbList/' . (int) $breadcrumbs->id];
+            $webPageSchema['breadcrumb'] = ['@id' => $breadcrumb['@id']];
+        } catch (UnknownAssetException $e) {
+            // No Breadcrumb Schema found, so we don't add it
```
```suggestion
            $webPageSchema['breadcrumbs'] = ['@id' => $breadcrumbs['@id']];
        } catch (UnknownAssetException $e) {
            // No Breadcrumbs Schema found, so we don't add it
```

HLeithner commented:
I have tested this item :white_check_mark: successfully on 982cdc31252128290ae87575b2894a6e463cd191<br /><br />Now it works for me.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/45720">issues.joomla.org/tracker/joomla-cms/45720</a>.</sub>

richard67 committed:
Merge branch '5.3-dev' into 5.3/schema-listitem

bembelimen committed:
Update modules/mod_breadcrumbs/tmpl/default.php

bembelimen committed:
Apply suggestions from code review

bembelimen committed:
Merge branch '5.3-dev' into 5.3/schema-listitem

brianteeman commented on plugins/system/schemaorg/src/Extension/Schemaorg.php:
```diff
@@ -383,11 +387,30 @@ public function onBeforeCompileHead(): void
+                trigger_deprecation(
+                    'joomla/schemaorg',
+                    '5.4',
+                    'The inline.mod_breadcrumbs-schemaorg asset name is deprecated. Please use the generic inline.breadcrumbs-schemaorg asset name instead.'
```
shouldnt there be a version when it will be removed?

bembelimen replied:
Thanks for the objection. The trigger calls normally do not contain the version when it's removed.

brianteeman replied:
it wasnt an objection it was a question based on the use of trigger_deprecation in libraries/joomla 

Example
https://github.com/joomla-framework/string/blob/0b3d33564db389e27346f7e275c694897c939434/src/Inflector.php#L105-L112

bembelimen committed:
Merge branch '5.3-dev' into 5.3/schema-listitem

heelc29 commented on plugins/system/schemaorg/src/Extension/Schemaorg.php:
```diff
@@ -383,11 +387,30 @@ public function onBeforeCompileHead(): void
 
-        if (!empty($breadcrumbs->id)) {
-            $webPageSchema['breadcrumb'] = ['@id' => $domain . '#/schema/BreadcrumbList/' . (int) $breadcrumbs->id];
+            $webPageSchema['breadcrumbs'] = ['@id' => $breadcrumbs['@id']];
```
@bembelimen I think this change is not correct
<img width="847" height="395" alt="image" src="https://github.com/user-attachments/assets/9ca8eb1b-cbd8-4524-b7b6-b9195995045f" />

heelc29 replied:
<img width="1473" height="538" alt="image" src="https://github.com/user-attachments/assets/3982c20c-2a76-4716-8937-a32a0e2bd13b" />
